### PR TITLE
feat: evaluate literals and variables in calling thread in `ParYield`

### PR DESCRIPTION
**Pull request by jaschdoc**
Suppose we have
```flix
let b = 1;
par (
        a <- b;
        c <- 2;
        e <- exp1;
        f <- exp2
    ) yield exp3
```

This PR changes `ParYield` to no longer spawn threads for `a` and `c` since these are just a variable lookup and a literal respectively.

Related to https://github.com/flix/flix/issues/4609

### CODE REVIEW

File: main/test/flix/Test.Exp.ParYield.flix [hunk 1]
```diff
@@ -66,4 +66,67 @@ namespace Test/Exp/ParYield {
     def testParYield15(): Bool =
         (par (a <- 5; b <- (let a = 1; a); c <- 6) yield a + b + c) == 12
 
+    @test
+    def testParYield16(): Int32 =
+        let c = 2;
+        par (
+            a <- 1;              // Simple variable
```
Comment (magnus-madsen) [inline main/test/flix/Test.Exp.ParYield.flix:73]: Constant

Comment (magnus-madsen) [reply main/test/flix/Test.Exp.ParYield.flix:73]: (Also below :))

File: main/src/ca/uwaterloo/flix/language/phase/Lowering.scala [hunk 2]
```diff
@@ -1641,6 +1648,31 @@ object Lowering {
...
+  private def spawnable(frag: LoweredAst.ParYieldFragment): Boolean = !isSimple(frag.exp)
+
+  /**
+    * Returns `true` if `exp0` is either a literal or a variable.
+    */
+  @tailrec
+  private def isSimple(exp0: LoweredAst.Expression): Boolean = exp0 match {
+    case LoweredAst.Expression.Var(_, _, _) => true
+    case LoweredAst.Expression.Cst(_: Ast.Constant, _, _) => true
+    case LoweredAst.Expression.Scope(_, _, exp, _, _, _, _) => isSimple(exp)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Lowering.scala:1672]: Lets not do this. For regions we leave it alone.

File: main/src/ca/uwaterloo/flix/language/phase/Lowering.scala [hunk 3]
```diff
@@ -1641,6 +1648,31 @@ object Lowering {
...
+  private def mkParYieldMainThread(exps: List[LoweredAst.ParYieldFragment], waitExps: LoweredAst.Expression): LoweredAst.Expression = {
+    exps.foldRight(waitExps) {
+      case (exp, acc) => mkLetMatch(exp.pat, exp.exp, acc)
+    }
+  }
+
+  /**
+    * Returns `true` if the ParYield fragment should be spawned in a thread. Wrapper for `isSimple`.
+    */
+  private def spawnable(frag: LoweredAst.ParYieldFragment): Boolean = !isSimple(frag.exp)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Lowering.scala:1663]: isSpawnable?

File: main/src/ca/uwaterloo/flix/language/phase/Lowering.scala [hunk 4]
```diff
@@ -1641,6 +1648,31 @@ object Lowering {
...
+
+  /**
+    * Returns `true` if `exp0` is either a literal or a variable.
+    */
+  @tailrec
+  private def isSimple(exp0: LoweredAst.Expression): Boolean = exp0 match {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Lowering.scala:1669]: isVarOrCst

File: main/src/ca/uwaterloo/flix/language/phase/Lowering.scala [hunk 5]
```diff
@@ -1641,6 +1648,31 @@ object Lowering {
     LoweredAst.Expression.Cast(blockExp, None, Some(Type.Pure), Some(Type.Empty), tpe, pur, eff, loc.asSynthetic)
   }
 
+  /**
+    * Returns the expression of a `ParYield` expression that should be evaluated in the calling thread.
+    */
+  private def mkParYieldMainThread(exps: List[LoweredAst.ParYieldFragment], waitExps: LoweredAst.Expression): LoweredAst.Expression = {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Lowering.scala:1654]: Not "Main" thread, maybe current thread?

Comment (jaschdoc) [reply main/src/ca/uwaterloo/flix/language/phase/Lowering.scala:1654]: Calling thread?

File: main/src/ca/uwaterloo/flix/language/phase/Lowering.scala (hunk 2 above)
Comment (jaschdoc) [reply main/src/ca/uwaterloo/flix/language/phase/Lowering.scala:1672]: Oh yes of course. For some reason I had this confused with a block exp i.e. `{ exp }`

File: main/src/ca/uwaterloo/flix/language/phase/Lowering.scala (hunk 5 above)
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/Lowering.scala:1654]: CurrentThread? :)

**Comment by jaschdoc** [comment]
Is there no GH action running for the latest commit? Tests are green locally

**Comment by magnus-madsen** [comment]
> Is there no GH action running for the latest commit? Tests are green locally

Sometimes it breaks, it should reset by itself.

**Comment by magnus-madsen** [comment]
Still stuck :S

**Pull request merged**